perf(platform-browser): disable styles of removed components instead of removing

mattlewis92 opened the pull request:
This commit changes the behaviour of `REMOVE_STYLES_ON_COMPONENT_DESTROY`.

Now, `style` and `link` nodes are disabled instead of removed from DOM which avoiding the browser having to reparse the stylesheet when the stylesheet is re-enabled.

This was previously reverted in 1640743b18e0b6a0e501855d5bc21414b2613319 but it seems like the observed perf regression in chromium has since been fixed in https://issues.chromium.org/issues/40911913, so it should be possible to reland this perf improvement

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

When all components using a stylesheet are destroyed, the stylesheet is removed from the DOM

Issue Number: 59651


## What is the new behavior?

The stylesheet is now marked as `disabled`, which improves performance the next time the stylesheet is enabled, as the browser will not have to incur the cost of reparsing the stylesheet when enabling it (although it will still trigger style recalculations)


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

When a component is destroyed, its associated `<style>` elements are now
disabled in the DOM rather than removed. When the same component is recreated, the styles
are re-enabled at their original position instead of being re-appended to the end of
`<head>`. This may change the CSS cascade order of component stylesheets, which could
affect styling if your application relies on the order in which component styles appear
in the DOM.

To migrate, avoid depending on the insertion order of Angular component stylesheets in
`<head>`. If you have styles that must take precedence over component styles, use more
specific CSS selectors or increase specificity rather than relying on source order.
Applications that do not rely on stylesheet ordering in `<head>` are not affected.

## Other information

alan-agius4 commented on packages/platform-browser/src/dom/shared_styles_host.ts:
```diff
@@ -25,8 +25,9 @@ const APP_ID_ATTRIBUTE_NAME = 'ng-app-id';
  */
 interface UsageRecord<T> {
-  elements: T[];
+  elements: Map</** Host */ Node, /** Style Node */ T>;
```
Why do we need a `Map` here?

mattlewis92 replied:
What about here when we remove the host?

https://github.com/angular/angular/blob/39bb0132a72b7d40262d4334c51c0a73a8cf9566/packages/platform-browser/src/dom/shared_styles_host.ts#L259-L261

alan-agius4 replied:
the `removeHost` is used used only in the `ShadowDomRenderer`, and in this case both `this.inline` and `this.external` will not be populated for the host that was added via `addHost`as `addStyles` and `removeStyles` are not used for shadow dom.

thePunderWoman reviewed (approved):
This looks good to me. I think since you're re-using existing codepaths, this shouldn't have any impact on the animations use case. So we should be good.

JeanMeche commented:
I'll take care of running a TGP over the weekend.

JeanMeche commented:
[TGP](https://fusion2.corp.google.com/presubmit/821373765/OCL:821373765:BASE:824266338:1761517999053:a8d86d32) is failing. `removeStyles` has some usages, because even though `ɵSharedStylesHost` is private its pulled by at least one project.

mattlewis92 commented:
@thePunderWoman was the internal usage of the `addStyles` api cleaned up yet to unblock this PR? Looks like the signature of `addStyles` is changing in #66048 as well, so I guess it will need to be cleaned up to unblock both PRs

thePunderWoman commented:
@mattlewis92 I'm unaware of the addStyles cleanup issue. Is there an open issue associated?

mattlewis92 committed:
perf(platform-browser): disable styles of removed components instead of removing

When a component is destroyed, instead of removing its style elements from
the DOM, we now disable them. This avoids expensive style recalculations
triggered by DOM removal while still preventing the styles from applying.
When the same component is re-created, the styles are re-enabled.

thePunderWoman commented:
@mattlewis92 Quick update: the reason for the delayed review is that @dgp1130 has been working on a style encapsulation PR for a while, and I was worried this would throw a wrench into his work. Once his stuff has landed, I can take a look.

JeanMeche commented:
We'll need a rebase after the latest changes we did on the SharedStylesHost.

mattlewis92 commented:
> We'll need a rebase after the latest changes we did on the SharedStylesHost.

Done!

thePunderWoman commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -599,20 +603,31 @@ class NoneEncapsulationDomRenderer extends DefaultDomRenderer2 {
+    // no-op. This is necessary because `BaseAnimationRenderer.destroy()` defers
+    // `delegate.destroy()` via `afterFlushAnimationsDone` + `queueMicrotask`.
+    // Without this, the deferred `disableStyles()` would find the *replacement*
+    // component's newly-added style record (same content key) and disable it.
```
Can we add a TODO to clean this up after the legacy animation package is removed?

mattlewis92 replied:
Good idea! Done in https://github.com/angular/angular/pull/64524/commits/3e3f6d49ed2e2af484bbbf585da0168aa33eb5b7

thePunderWoman reviewed (approved):
LGTM

mattlewis92 committed:
fixup! perf(platform-browser): disable styles of removed components instead of removing

Add disableStyles to ɵSharedStylesHost interface and deprecate removeStyles

mattlewis92 committed:
fixup! perf(platform-browser): disable styles of removed components instead of removing

Add TODO to clean up style-clearing workaround when legacy animations package is removed
